episodes/3-documentation.md

```diff
@@ -216,49 +215,148 @@ def fahrenheit_to_celsius(temp_f: float) -> float:
 The second set uses **docstrings or similar concepts**. Please compare the two
 (above and below):
 
-```py
+```python
 def fahrenheit_to_celsius(temp_f: float) -> float:
     """
     Converts a temperature in Fahrenheit to Celsius.
 
-    Parameters
-    ----------
-    temp_f : float
-        The temperature in Fahrenheit.
+    Args:
+        temp_f (float): The temperature in Fahrenheit.
 
-    Returns
-    -------
-    float
-        The temperature in Celsius.
+    Returns:
+        float: The temperature in Celsius.
     """
 
     temp_c = (temp_f - 32.0) * (5.0/9.0)
     return temp_c
 ```
 
+Unlike regular comments, docstrings are stored as an attribute (`__doc__`) of the object they document and can be extracted programmatically. 
+
 Docstrings are more powerful than comments:
 
-- Docstrings are automatically extracted when calling the help function.
-- Tools can generate documentation pages automatically from docstrings known as API documentation.
+- Docstrings are automatically extracted when calling the `help` function.
+- Tools (e.g. `mkdocs`, `sphinx`) can generate documentation pages automatically from docstrings known as API documentation.
 
-It is common to write docstrings for functions, classes, and modules.
+A good docstring should describe:
 
-TODO: Introduce how to write docstrings and show an example of API documentation.
-Good docstrings describe:
 - What the function does
 - What goes in (including the type of the input variables)
 - What goes out (including the return type)
 
+There are different docstrings styles, such as Google style, numpy style and reStructered style, with Google style being the most common.
+
+::::::::::::::::::::::::::::::::::::: challenge 
+
+## Exercise: Write a docstring 
+
+``` python
+def check_temperature_validity(temperature, unit):
+    abs_zero = {"C": -273.15, "F": -459.67, "K": 0}
+    if temperature < abs_zero[unit]:
+        return False
+    return True
+
+```
+:::::::::::::::::::::::: solution 
+
+## Solution
+
+Caveat: without specifying the expected data types of function arguments and return values it is unclear what these values should be! This is called _type hinting_. Lets add types and write the docstring.
+
+``` python
+def check_temperature_validity(temperature: float, unit: str) -> bool:
+    """
+    Checks if a temperature is valid for a given unit.
+
+    Args:
+        temperature (float): The temperature to check.
+        unit (str): The unit of the temperature. Must be "C", "F", or "K".
+
+    Returns:
+        bool: True if the temperature is valid, False otherwise.
+    """
+    abs_zero = {"C": -273.15, "F": -459.67, "K": 0}
+    if temperature < abs_zero[unit]:
+        return False
+    return True
+```
+:::::::::::::::::::::::: 
+
+:::::::::::::::::::::::::::::::::::::
+
+
+::::::::::::::::::::::::::::::::::::: challenge 
+## Exercise: Identify proper docstring
+
+Which of the following options is the correct function with a proper docstring?
```

Comment:
I like this exercise! I think it can be done quickly, but still brings across the major points of writing good docstrings.